### PR TITLE
feat: Inclusion of an invoice notes text field

### PR DESCRIPTION
**Pull request by pedroCloudoki**

### CODE REVIEW

File: lib/pages/Billing/Billing.tsx [hunk 1]
```diff
@@ -485,6 +502,61 @@ const Billing: React.FC<BillingProps> = ({
...
+            variant="outlined"
+          />
+        </Box>
+
+        <Box clone mt={3}>
+          <Button
+            color="primary"
+            disabled={
+              (userInvoiceNotes && userInvoiceNotes === invoiceNotes) ||
+              (!userInvoiceNotes && userInvoiceNotes === invoiceNotes)
```
Comment (s1moe2) [inline lib/pages/Billing/Billing.tsx:549]: Can be simplified (and clearer)

File: lib/pages/Billing/sagas.ts [hunk 2]
```diff
@@ -109,6 +115,48 @@ export function* getSubscriptionPlansActionSaga(
...
+        'content-type': 'application/json',
+      },
+      data: {
+        invoiceNotes: action.invoiceNotes,
+      },
+    })
+
+    yield put(setUserInvoiceNotesActionSuccess(response.data.invoiceNotes))
+  } catch (error) {
+    console.log("Error setting the user's invoice notes.")
```
Comment (a31859) [inline lib/pages/Billing/sagas.ts:156]: give user error feedback instead of `console.log`

File: lib/pages/Billing/sagas.ts [hunk 3]
```diff
@@ -109,6 +115,48 @@ export function* getSubscriptionPlansActionSaga(
...
+  }
+}
+
+export function* setUserInvoiceNotesActionSaga(
+  action: SetUserInvoiceNotesAction
+) {
+  try {
+    const setUserInvoiceNotesActionUrl = `${BILLING_API_URL}/users/${action.userID}/invoice-notes`
+
+    const response = yield call(request, {
```
Comment (a31859) [inline lib/pages/Billing/sagas.ts:143]: Add response type

File: lib/pages/Billing/sagas.ts [hunk 4]
```diff
@@ -109,6 +115,48 @@ export function* getSubscriptionPlansActionSaga(
   }
 }
 
+export function* getUserInvoiceNotesActionSaga(
+  action: GetUserInvoiceNotesAction
+) {
+  try {
+    const getUserInvoiceNotesActionUrl = `${BILLING_API_URL}/users/${action.userID}/invoice-notes`
+
+    const response = yield call(request, {
```
Comment (a31859) [inline lib/pages/Billing/sagas.ts:124]: add response type

File: lib/pages/Billing/sagas.ts [hunk 5]
```diff
@@ -109,6 +115,48 @@ export function* getSubscriptionPlansActionSaga(
...
+      url: getUserInvoiceNotesActionUrl,
+      method: 'GET',
+    })
+
+    const invoiceNotes = response.data.invoiceNotes
```
Comment (a31859) [inline lib/pages/Billing/sagas.ts:129]: is this variable needed?

File: lib/pages/Billing/Billing.tsx [hunk 6]
```diff
@@ -485,6 +502,61 @@ const Billing: React.FC<BillingProps> = ({
           </>
         )}
 
+        {/* 'Invoice notes' section */}
+
+        <Box clone mb={1.5} mt={5}>
+          <Typography variant="h3">
+            {t('invoiceNotes.invoiceNotesTitle')}
```
Comment (a31859) [inline lib/pages/Billing/Billing.tsx:509]: `invoiceNotes.invoiceNotesTitle` avoid double naming. `invoiceNotes.title`

File: lib/pages/Billing/Billing.tsx [hunk 7]
```diff
@@ -485,6 +502,61 @@ const Billing: React.FC<BillingProps> = ({
...
+          </Typography>
+        </Box>
+
+        <Box clone mb={3}>
+          <Typography variant="body1" color="textSecondary">
+            {t('invoiceNotes.invoiceNotesSubtitle')}
```
Comment (a31859) [inline lib/pages/Billing/Billing.tsx:515]: double naming

File: lib/pages/Billing/Billing.tsx [hunk 8]
```diff
@@ -485,6 +502,61 @@ const Billing: React.FC<BillingProps> = ({
...
+            width: '100%',
+          }}
+        >
+          <TextField
+            className={classes.invoiceNotesTextField}
+            fullWidth
+            InputLabelProps={{
+              shrink: true,
+            }}
+            label={t('invoiceNotes.invoiceNotesTextFieldLabel')}
```
Comment (a31859) [inline lib/pages/Billing/Billing.tsx:533]: double naming

File: lib/pages/Billing/Billing.tsx [hunk 9]
```diff
@@ -485,6 +502,61 @@ const Billing: React.FC<BillingProps> = ({
...
+            disabled={
+              (userInvoiceNotes && userInvoiceNotes === invoiceNotes) ||
+              (!userInvoiceNotes && userInvoiceNotes === invoiceNotes)
+            }
+            disableElevation
+            onClick={() => setUserInvoiceNotesAction(user.id, userInvoiceNotes)}
+            size="large"
+            variant="contained"
+          >
+            {t('invoiceNotes.saveInvoiceNoteButtonLabel')}
```
Comment (a31859) [inline lib/pages/Billing/Billing.tsx:556]: double naming

File: lib/pages/Billing/ducks.ts [hunk 10]
```diff
@@ -22,6 +22,7 @@ const initialState: BillingStore = {
   },
   creditPacks: [],
   subscriptions: [],
+  invoiceNotes: '',
```
Comment (a31859) [inline lib/pages/Billing/ducks.ts:25]: If `invoiceNotes` is a single string think of using `invoiceNote` instead plural assumes multiple notes but seems to be just one.

File: lib/pages/Billing/sagas.ts [hunk 11]
```diff
@@ -109,6 +115,48 @@ export function* getSubscriptionPlansActionSaga(
...
+
+    yield put(getUserInvoiceNotesActionSuccess(invoiceNotes))
+  } catch (error) {
+    console.log("Error fetching the user's invoice notes.")
```
Comment (a31859) [inline lib/pages/Billing/sagas.ts:133]: Feedback to user instead of console.log

File: lib/pages/Billing/types.ts [hunk 12]
```diff
@@ -125,6 +125,10 @@ export interface BillingProps {
   user: User
 }
 
+export interface GetAndSetUserInvoiceNoteResponse {
```
Comment (a31859) [inline lib/pages/Billing/types.ts:128]: `GetAndSetUserInvoiceNoteResponse` -> `InvoiceNotesResponse`

**Review by a31859 — CHANGES_REQUESTED** [review]

**Pull request merged**